src/engine_empty.cc

```diff
@@ -73,6 +73,16 @@ class EmptyEngine : public IEngine {
     // simply print information into the tracker
     utils::Printf("%s", msg.c_str());
   }
+  virtual int SetCache(const std::string &key, const void* buf, const size_t buflen) {
```

Comment:
if this thing is only for cases that allreduce worker are having different allreduce operations in different stage, I would suggest just remove it 

let's only support the simplest case for now, from https://github.com/dmlc/xgboost/pull/4636/files, there is no calling of setcache for approx algorithm, that means we can rely on the original checkpoint mechanism for the default algorithm

Reply:
ignore this, as I said in xgb repo, it's very complicated actually